server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsPhase.java

```diff
@@ -35,7 +35,7 @@ public InnerHitsPhase(FetchPhase fetchPhase) {
 
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext searchContext) {
-        if (searchContext.innerHits() == null) {
+        if (searchContext.innerHits() == null || searchContext.innerHits().getInnerHits().isEmpty()) {
```

Comment:
Not really related but I saw we were running inner hits for every doc even if none were asked for because by default we return an empty list here, not null.